# 기본 설정

In [4]:
import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
%matplotlib inline
from wordcloud import WordCloud, STOPWORDS as stopwords
from PIL import Image, ImageFilter
from wordcloud import ImageColorGenerator
from matplotlib import font_manager, rc
from collections import Counter
import konlpy
from konlpy.tag import Mecab
import time
import seaborn as sns
import matplotlib.pyplot as plt
from random import shuffle
import json
from tqdm.notebook import tqdm
tqdm.pandas()
import gensim 
from gensim.models import Word2Vec, FastText, Doc2Vec
import fasttext
from bokeh.plotting import figure, output_file, save
from bokeh.io import curdoc, show
from bokeh.transform import factor_cmap, factor_mark
from bokeh.models import ColumnDataSource, Grid, LinearAxis, Plot, Text
# from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import HoverTool, ColumnDataSource, value
from sklearn.feature_extraction.text import TfidfVectorizer
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import HoverTool, ColumnDataSource, value
output_notebook()
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tqdm/std.py:666: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  from pandas import Panel


Loading BokehJS ...

In [9]:
bills = pd.read_csv('/Users/junghyunwoo/혀누에-의한-혀누를-위한-혀누의/공부공부공부공부/대학 이후/코딩 관련/개인프로젝트_국회의원 추천 (20대)/1차 데이터 및 jupyter/bills_24839_all_cleansed.csv')
senators = pd.read_csv('/Users/junghyunwoo/혀누에-의한-혀누를-위한-혀누의/공부공부공부공부/대학 이후/코딩 관련/개인프로젝트_국회의원 추천 (20대)/1차 데이터 및 jupyter/(원본) assembly_memebers_295.csv')

In [7]:
senators_w_clusters = pd.read_csv("/Users/junghyunwoo/혀누에-의한-혀누를-위한-혀누의/공부공부공부공부/대학 이후/코딩 관련/개인프로젝트_국회의원 추천 (20대)/1차 데이터 및 jupyter/senator with cluster_13 ver21.0.csv")
senators_w_clusters.rename(columns={'index':'한글이름'}, inplace=True)

In [10]:
senators['한글이름'] = senators['이름'].str.split(' ').str[0]
senators['한자이름'] = senators['이름'].str.split(' ').str[1]

In [11]:
bills = bills[bills['발의자'] != '위원장']
bills = bills[bills['발의자'] != '정부'] 
bills = bills.dropna()

# methods

In [16]:
def count_bills(name, bills_df):
    df_temp = bills_df[bills_df['발의자'].str.contains(name)]
    bills_deleted = df_temp[(df_temp['상태']=='폐기') | (df_temp['상태']=='철회')].shape[0]
    df_temp_save = df_temp[~(df_temp['상태']=='폐기') & ~(df_temp['상태']=='철회')]
    bills_일부개정법률안 = df_temp_save[df_temp_save['법안명'].str.contains('일부개정법률안')].shape[0]
    bills_전부개정법률안 = df_temp_save[df_temp_save['법안명'].str.contains('전부개정법률안')].shape[0]
    bills_제정법안 = df_temp_save[(~df_temp_save['법안명'].str.contains('일부개정법률안'))& (~df_temp_save['법안명'].str.contains('전부개정법률안'))].shape[0]
    dict_ = {'일부개정법률안 비율': str(round((bills_일부개정법률안/df_temp.shape[0])*100,2))+'%',
            '전부개정법률안 비율': str(round((bills_전부개정법률안/df_temp.shape[0])*100,2))+'%',
            '제정법률안 비율': str(round((bills_제정법안/df_temp.shape[0])*100,2))+"%",
            '폐기법률안 비율': str(round((bills_deleted/df_temp.shape[0])*100,2))+'%'}
    return dict_

def count_bills_all(bills_df):
    df_temp = bills_df
    bills_deleted = df_temp[(df_temp['상태']=='폐기') | (df_temp['상태']=='철회')].shape[0]
    df_temp_save = df_temp[~(df_temp['상태']=='폐기') & ~(df_temp['상태']=='철회')]
    bills_일부개정법률안 = df_temp_save[df_temp_save['법안명'].str.contains('일부개정법률안')].shape[0]
    bills_전부개정법률안 = df_temp_save[df_temp_save['법안명'].str.contains('전부개정법률안')].shape[0]
    bills_제정법안 = df_temp_save[(~df_temp_save['법안명'].str.contains('일부개정법률안'))& (~df_temp_save['법안명'].str.contains('전부개정법률안'))].shape[0]
    dict_ = {'일부개정법률안 비율': str(round((bills_일부개정법률안/df_temp.shape[0])*100,2))+'%',
            '전부개정법률안 비율': str(round((bills_전부개정법률안/df_temp.shape[0])*100,2))+'%',
            '제정법률안 비율': str(round((bills_제정법안/df_temp.shape[0])*100,2))+"%",
            '폐기법률안 비율': str(round((bills_deleted/df_temp.shape[0])*100,2))+'%'}
    return dict_


def get_quantile(senator):
    result = result
    return result

# 전체 distribution 분포 살펴보기

In [17]:
dict_bills_portion = {}
for senator in senators['한글이름']:
    dict_bills_portion[senator] = count_bills(senator, bills)

In [18]:
df_bills_portion = pd.DataFrame(dict_bills_portion).T

In [19]:
df_bills_portion['일부개정법률안 비율'] = df_bills_portion['일부개정법률안 비율'].str[:-1]
df_bills_portion['전부개정법률안 비율'] = df_bills_portion['전부개정법률안 비율'].str[:-1]
df_bills_portion['폐기법률안 비율'] = df_bills_portion['폐기법률안 비율'].str[:-1]
df_bills_portion['제정법률안 비율'] = df_bills_portion['제정법률안 비율'].str[:-1]

In [20]:
df_bills_portion['일부개정법률안 비율'] = df_bills_portion['일부개정법률안 비율'].astype(float)
df_bills_portion['전부개정법률안 비율'] = df_bills_portion['전부개정법률안 비율'].astype(float)
df_bills_portion['폐기법률안 비율'] = df_bills_portion['폐기법률안 비율'].astype(float)
df_bills_portion['제정법률안 비율'] = df_bills_portion['제정법률안 비율'].astype(float)

In [21]:
df_bills_portion['제정및전부개정법률안 비율'] = df_bills_portion['제정법률안 비율'] + df_bills_portion['전부개정법률안 비율']

# 상위 % 잘라낸 뒤 추천하기

In [22]:
df_bills_portion.reset_index(inplace=True)

In [23]:
senators = senators.merge(df_bills_portion, left_on="한글이름", right_on="index", how="left")
senators = senators.drop(columns=['index'])

In [24]:
df_temp = senators[['이름', '한글이름','제정및전부개정법률안 비율']].sort_values(['제정및전부개정법률안 비율'])

In [25]:
df_temp['순위'] = range(1, df_temp.shape[0]+1)

In [26]:
df_temp['percentile'] = round(100-(df_temp['순위']/295) * 100, 5)

In [27]:
df_temp['percentile'] = df_temp['percentile'].astype(str)
df_temp['percentile'] = df_temp['percentile'] + '%'
df_temp

이름 한글이름  제정및전부개정법률안 비율   순위 percentile
284  허윤정 許允貞  허윤정           0.00    1  99.66102%
245  정은혜 鄭恩惠  정은혜           0.00    2  99.32203%
248  정점식 鄭点植  정점식           2.63    3  98.98305%
271  최도자 崔道子  최도자           3.89    4  98.64407%
224  임재훈 林哉勳  임재훈           4.32    5  98.30508%
14   권칠승 權七勝  권칠승           4.38    6   97.9661%
181  이규희 李揆熙  이규희           4.41    7  97.62712%
220  이후삼 李厚三  이후삼           4.42    8  97.28814%
15   금태섭 琴泰燮  금태섭           4.44    9  96.94915%
136  신용현 申容賢  신용현           4.49   10  96.61017%
190  이언주 李彦周  이언주           4.51   11  96.27119%
173  윤일규 尹一逵  윤일규           4.59   12   95.9322%
142  안규백 安圭佰  안규백           4.60   13  95.59322%
274  최인호 崔仁昊  최인호           4.81   14  95.25424%
182  이동섭 李銅燮  이동섭           4.84   15  94.91525%
74   김해영 金海永  김해영           4.92   16  94.57627%
96   박선숙 朴仙淑  박선숙           4.92   17  94.23729%
267  채이배 蔡利培  채이배           4.99   18  93.89831%
233  전혜숙 全惠淑  전혜숙           5.02   19  93.55932%
228  장정숙 張貞淑  장정숙           5.07   20  93.22034%
65   김중로 金中魯  김중로           5.08   21  92.88136%
33   김삼화 金三和  김삼화           5.09   22  92.54237%
202  이종걸 李鍾杰  이종걸           5.16   23  92.20339%
213  이태규 李泰珪  이태규           5.20   24  91.86441%
50   김영진 金榮鎭  김영진           5.21   25  91.52542%
208  이찬열 李燦烈  이찬열           5.21   26  91.18644%
125  송갑석 宋甲錫  송갑석           5.21   27  90.84746%
89   박광온 朴洸瑥  박광온           5.22   28  90.50847%
137  신창현 申昌賢  신창현           5.34   29  90.16949%
18   김관영 金寬永  김관영           5.42   30  89.83051%
214  이학영 李學永  이학영           5.44   31  89.49153%
110  박찬대 朴贊大  박찬대           5.48   32  89.15254%
230  전재수 田載秀  전재수           5.48   33  88.81356%
292  황주홍 黃柱洪  황주홍           5.48   34  88.47458%
276  추경호 秋慶鎬  추경호           5.61   35  88.13559%
262  주승용 朱昇鎔  주승용           5.61   36  87.79661%
114  백혜련 白惠蓮  백혜련           5.66   37  87.45763%
8    고용진 高榕禛  고용진           5.70   38  87.11864%
38   김성식 金成植  김성식           5.73   39  86.77966%
275  최재성 崔宰誠  최재성           5.88   40  86.44068%
82   맹성규 孟聖奎  맹성규           5.92   41  86.10169%
187  이상헌 李相憲  이상헌           5.93   42  85.76271%
132  신경민 辛京珉  신경민           5.96   43  85.42373%
167  윤관석 尹官石  윤관석           5.99   44  85.08475%
93   박명재 朴明在  박명재           6.00   45  84.74576%
178  윤호중 尹昊重  윤호중           6.01   46  84.40678%
83   문진국 文鎭國  문진국           6.02   47   84.0678%
104  박재호 朴在昊  박재호           6.04   48  83.72881%
108  박주현 朴珠賢  박주현           6.05   49  83.38983%
258  조응천 趙應天  조응천           6.06   50  83.05085%
57   김정재 金汀才  김정재           6.06   51  82.71186%
254  조경태 趙慶泰  조경태           6.09   52  82.37288%
232  전현희 全賢姬  전현희           6.11   53   82.0339%
49   김영주 金榮珠  김영주           6.11   54  81.69492%
130  송옥주 宋玉珠  송옥주           6.12   55  81.35593%
239  정세균 丁世均  정세균           6.13   56  81.01695%
172  윤영일 尹泳一  윤영일           6.14   57  80.67797%
64   김경진 金京鎭  김경진           6.15   58  80.33898%
58   김정호 金禎鎬  김정호           6.18   59      80.0%
121  성일종 成一鍾  성일종           6.22   60  79.66102%
56   김정우 金政祐  김정우           6.26   61  79.32203%
246  정인화 鄭仁和  정인화           6.30   62  78.98305%
91   박덕흠 朴德欽  박덕흠           6.32   63  78.64407%
124  손혜원 孫惠園  손혜원           6.40   64  78.30508%
23   김도읍 金度邑  김도읍           6.42   65   77.9661%
194  이원욱 李元旭  이원욱           6.49   66  77.62712%
20   김광수 金光守  김광수           6.50   67  77.28814%
212  이춘석 李春錫  이춘석           6.51   68  76.94915%
30   김병기 金炳基  김병기           6.52   69  76.61017%
171  윤영석 尹永碩  윤영석           6.55   70  76.27119%
184  이명수 李明洙  이명수           6.57   71   75.9322%
45   김수민 金秀玟  김수민           6.57   72  75.59322%
87   민홍철 閔洪喆  민홍철           6.57   73  75.25424%
145  안호영 安浩永  안호영           6.60   74  74.91525%
247  정재호 鄭在浩  정재호           6.60   75  74.57627%
29   김병관 金炳官  김병관           6.62   76  74.23729%
205  이종배 李鍾培  이종배           6.64   77  73.89831%
235  정갑윤 鄭甲潤  정갑윤           6.71   78  73.55932%
159  유동수 柳東秀  유동수           6.72   79  73.22034%
259  조정식 趙正湜  조정식           6.76   80  72.88136%
4    강창일 姜昌一  강창일           6.77   81  72

# Calculate senators' bills index

In [29]:
def count_senators(data):
    data = [x for x in data if '/' in x]
    return len(data)

In [30]:
dict_count_biils = {}
for row in senators.itertuples():
    dict_temp = {}
    dict_temp['한글이름'] = row[-7]
    
    # Get the dataframe
    df_temp = bills[bills['발의자'].str.contains(row[-7])]
    bills_전부개정법률안 = df_temp[df_temp['법안명'].str.contains('전부개정법률안')]
    bills_제정법안 = df_temp[(~df_temp['법안명'].str.contains('일부개정법률안'))& (~df_temp['법안명'].str.contains('전부개정법률안'))]
    bills_total = pd.concat([bills_제정법안, bills_전부개정법률안])
    
    # count mean senators proposing 
    bills_total['발의자 수'] = bills_total['발의자'].str.split(' ').map(count_senators)
    
    # save in dictionary
    dict_temp['전체 발의법안 수'] = bills_total.shape[0]
    if dict_temp['전체 발의법안 수'] == 0:
        dict_temp['평균 발의자 수'] = 0.000000000000009
    else:    
        dict_temp['평균 발의자 수'] = round(bills_total['발의자 수'].mean())

    dict_count_biils[row[-6]] = dict_temp

In [31]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

In [32]:
df_result = pd.DataFrame(dict_count_biils).T.reset_index().rename(columns={'index':'한자이름'})
df_result['법안수/평균 발의자 수'] = df_result['전체 발의법안 수'] / df_result['평균 발의자 수'] 
df_result.sort_values(by='법안수/평균 발의자 수')
df_result['bills_index'] = df_result['법안수/평균 발의자 수'].astype(float)
df_result['bills_index_scaled'] = scaler.fit_transform(df_result[['bills_index']])
df_result

한자이름 한글이름 전체 발의법안 수 평균 발의자 수 법안수/평균 발의자 수  bills_index  bills_index_scaled
0    姜吉夫  강길부        23       67     0.343284     0.343284            0.031618
1    姜炳遠  강병원        94       38      2.47368     2.473684            0.227839
2    姜錫振  강석진        61       47      1.29787     1.297872            0.119541
3    姜碩鎬  강석호        63       47      1.34043     1.340426            0.123460
4    姜昌一  강창일        78       38      2.05263     2.052632            0.189058
5    姜孝祥  강효상        38       63     0.603175     0.603175            0.055556
6    姜勳植  강훈식       116       34      3.41176     3.411765            0.314241
7    慶大秀  경대수        67       43      1.55814     1.558140            0.143513
8    高榕禛  고용진        75       34      2.20588     2.205882            0.203173
9    郭大勳  곽대훈        74       46       1.6087     1.608696            0.148169
10   郭尙道  곽상도        35       69     0.507246     0.507246            0.046720
11   權美赫  권미혁        94       37      2.54054     2.540541            0.233997
12   權性東  권성동        54       53      1.01887     1.018868            0.093843
13   權垠希  권은희        41       29      1.41379     1.413793            0.130218
14   權七勝  권칠승        80       36      2.22222     2.222222            0.204678
15   琴泰燮  금태섭        38       33      1.15152     1.151515            0.106061
16   奇東旻  기동민       130       33      3.93939     3.939394            0.362839
17   金炅俠  김경협       103       35      2.94286     2.942857            0.271053
18   金寬永  김관영        91       32      2.84375     2.843750            0.261924
19   金光琳  김광림        47       58     0.810345     0.810345            0.074637
20   金光守  김광수        88       27      3.25926     3.259259            0.300195
21   金奎煥  김규환        74       43      1.72093     1.720930            0.158507
22   金起善  김기선        50       62     0.806452     0.806452            0.074278
23   金度邑  김도읍        65       45      1.44444     1.444444            0.133041
24   金東喆  김동철        67       35      1.91429     1.914286            0.176316
25   金斗官  김두관        59       49      1.20408     1.204082            0.110902
26   金明淵  김명연        59       47      1.25532     1.255319            0.115622
27   金武星  김무성        40       71      0.56338     0.563380            0.051890
28   金敏基  김민기        90       38      2.36842     2.368421            0.218144
29   金炳官  김병관        60       44      1.36364     1.363636            0.125598
30   金炳基  김병기       105       34      3.08824     3.088235            0.284443
31   金炳旭  김병욱       106       34      3.11765     3.117647            0.287152
32   金富謙  김부겸        35       44     0.795455     0.795455            0.073266
33   金三和  김삼화        90       25          3.6     3.600000            0.331579
34   金相勳  김상훈        66       49      1.34694     1.346939            0.124060
35   金相姬  김상희       151       31      4.87097     4.870968            0.448642
36   金碩基  김석기        89       37      2.40541     2.405405            0.221550
37   金善東  김선동        56       48      1.16667     1.166667            0.107456
38   金成植  김성식        12       62     0.193548     0.193548            0.017827
39   金成願  김성원       100       35      2.85714     2.857143            0.263158
40   金盛贊  김성찬        95       37      2.56757     2.567568            0.236486
41   金成泰  김성태        83       44      1.88636     1.886364            0.173744
42   金聖泰  김성태        83       44      1.88636     1.886364            0.173744
43   金星煥  김성환         9       42     0.214286     0.214286            0.019737
44   金世淵  김세연        64       48      1.33333     1.333333            0.122807
45   金秀玟  김수민        81       24        3.375     3.375000            0.310855
46   金順禮  김순례        52       51      1.01961     1.019608            0.093911
47   金承禧  김승희        90       35      2.57143     2.571429            0.236842
48   金榮宇  김영우        51       52     0.980769     0.980769            0.090334
49   金榮珠  김영주        44       44            1     1.00

In [33]:
bills_index = df_result.copy()
bills_index

한자이름 한글이름 전체 발의법안 수 평균 발의자 수 법안수/평균 발의자 수  bills_index  bills_index_scaled
0    姜吉夫  강길부        23       67     0.343284     0.343284            0.031618
1    姜炳遠  강병원        94       38      2.47368     2.473684            0.227839
2    姜錫振  강석진        61       47      1.29787     1.297872            0.119541
3    姜碩鎬  강석호        63       47      1.34043     1.340426            0.123460
4    姜昌一  강창일        78       38      2.05263     2.052632            0.189058
5    姜孝祥  강효상        38       63     0.603175     0.603175            0.055556
6    姜勳植  강훈식       116       34      3.41176     3.411765            0.314241
7    慶大秀  경대수        67       43      1.55814     1.558140            0.143513
8    高榕禛  고용진        75       34      2.20588     2.205882            0.203173
9    郭大勳  곽대훈        74       46       1.6087     1.608696            0.148169
10   郭尙道  곽상도        35       69     0.507246     0.507246            0.046720
11   權美赫  권미혁        94       37      2.54054     2.540541            0.233997
12   權性東  권성동        54       53      1.01887     1.018868            0.093843
13   權垠希  권은희        41       29      1.41379     1.413793            0.130218
14   權七勝  권칠승        80       36      2.22222     2.222222            0.204678
15   琴泰燮  금태섭        38       33      1.15152     1.151515            0.106061
16   奇東旻  기동민       130       33      3.93939     3.939394            0.362839
17   金炅俠  김경협       103       35      2.94286     2.942857            0.271053
18   金寬永  김관영        91       32      2.84375     2.843750            0.261924
19   金光琳  김광림        47       58     0.810345     0.810345            0.074637
20   金光守  김광수        88       27      3.25926     3.259259            0.300195
21   金奎煥  김규환        74       43      1.72093     1.720930            0.158507
22   金起善  김기선        50       62     0.806452     0.806452            0.074278
23   金度邑  김도읍        65       45      1.44444     1.444444            0.133041
24   金東喆  김동철        67       35      1.91429     1.914286            0.176316
25   金斗官  김두관        59       49      1.20408     1.204082            0.110902
26   金明淵  김명연        59       47      1.25532     1.255319            0.115622
27   金武星  김무성        40       71      0.56338     0.563380            0.051890
28   金敏基  김민기        90       38      2.36842     2.368421            0.218144
29   金炳官  김병관        60       44      1.36364     1.363636            0.125598
30   金炳基  김병기       105       34      3.08824     3.088235            0.284443
31   金炳旭  김병욱       106       34      3.11765     3.117647            0.287152
32   金富謙  김부겸        35       44     0.795455     0.795455            0.073266
33   金三和  김삼화        90       25          3.6     3.600000            0.331579
34   金相勳  김상훈        66       49      1.34694     1.346939            0.124060
35   金相姬  김상희       151       31      4.87097     4.870968            0.448642
36   金碩基  김석기        89       37      2.40541     2.405405            0.221550
37   金善東  김선동        56       48      1.16667     1.166667            0.107456
38   金成植  김성식        12       62     0.193548     0.193548            0.017827
39   金成願  김성원       100       35      2.85714     2.857143            0.263158
40   金盛贊  김성찬        95       37      2.56757     2.567568            0.236486
41   金成泰  김성태        83       44      1.88636     1.886364            0.173744
42   金聖泰  김성태        83       44      1.88636     1.886364            0.173744
43   金星煥  김성환         9       42     0.214286     0.214286            0.019737
44   金世淵  김세연        64       48      1.33333     1.333333            0.122807
45   金秀玟  김수민        81       24        3.375     3.375000            0.310855
46   金順禮  김순례        52       51      1.01961     1.019608            0.093911
47   金承禧  김승희        90       35      2.57143     2.571429            0.236842
48   金榮宇  김영우        51       52     0.980769     0.980769            0.090334
49   金榮珠  김영주        44       44            1     1.00

# Compare bills index within clusters

In [35]:
bills_index

한자이름 한글이름 전체 발의법안 수 평균 발의자 수 법안수/평균 발의자 수  bills_index  bills_index_scaled
0    姜吉夫  강길부        23       67     0.343284     0.343284            0.031618
1    姜炳遠  강병원        94       38      2.47368     2.473684            0.227839
2    姜錫振  강석진        61       47      1.29787     1.297872            0.119541
3    姜碩鎬  강석호        63       47      1.34043     1.340426            0.123460
4    姜昌一  강창일        78       38      2.05263     2.052632            0.189058
5    姜孝祥  강효상        38       63     0.603175     0.603175            0.055556
6    姜勳植  강훈식       116       34      3.41176     3.411765            0.314241
7    慶大秀  경대수        67       43      1.55814     1.558140            0.143513
8    高榕禛  고용진        75       34      2.20588     2.205882            0.203173
9    郭大勳  곽대훈        74       46       1.6087     1.608696            0.148169
10   郭尙道  곽상도        35       69     0.507246     0.507246            0.046720
11   權美赫  권미혁        94       37      2.54054     2.540541            0.233997
12   權性東  권성동        54       53      1.01887     1.018868            0.093843
13   權垠希  권은희        41       29      1.41379     1.413793            0.130218
14   權七勝  권칠승        80       36      2.22222     2.222222            0.204678
15   琴泰燮  금태섭        38       33      1.15152     1.151515            0.106061
16   奇東旻  기동민       130       33      3.93939     3.939394            0.362839
17   金炅俠  김경협       103       35      2.94286     2.942857            0.271053
18   金寬永  김관영        91       32      2.84375     2.843750            0.261924
19   金光琳  김광림        47       58     0.810345     0.810345            0.074637
20   金光守  김광수        88       27      3.25926     3.259259            0.300195
21   金奎煥  김규환        74       43      1.72093     1.720930            0.158507
22   金起善  김기선        50       62     0.806452     0.806452            0.074278
23   金度邑  김도읍        65       45      1.44444     1.444444            0.133041
24   金東喆  김동철        67       35      1.91429     1.914286            0.176316
25   金斗官  김두관        59       49      1.20408     1.204082            0.110902
26   金明淵  김명연        59       47      1.25532     1.255319            0.115622
27   金武星  김무성        40       71      0.56338     0.563380            0.051890
28   金敏基  김민기        90       38      2.36842     2.368421            0.218144
29   金炳官  김병관        60       44      1.36364     1.363636            0.125598
30   金炳基  김병기       105       34      3.08824     3.088235            0.284443
31   金炳旭  김병욱       106       34      3.11765     3.117647            0.287152
32   金富謙  김부겸        35       44     0.795455     0.795455            0.073266
33   金三和  김삼화        90       25          3.6     3.600000            0.331579
34   金相勳  김상훈        66       49      1.34694     1.346939            0.124060
35   金相姬  김상희       151       31      4.87097     4.870968            0.448642
36   金碩基  김석기        89       37      2.40541     2.405405            0.221550
37   金善東  김선동        56       48      1.16667     1.166667            0.107456
38   金成植  김성식        12       62     0.193548     0.193548            0.017827
39   金成願  김성원       100       35      2.85714     2.857143            0.263158
40   金盛贊  김성찬        95       37      2.56757     2.567568            0.236486
41   金成泰  김성태        83       44      1.88636     1.886364            0.173744
42   金聖泰  김성태        83       44      1.88636     1.886364            0.173744
43   金星煥  김성환         9       42     0.214286     0.214286            0.019737
44   金世淵  김세연        64       48      1.33333     1.333333            0.122807
45   金秀玟  김수민        81       24        3.375     3.375000            0.310855
46   金順禮  김순례        52       51      1.01961     1.019608            0.093911
47   金承禧  김승희        90       35      2.57143     2.571429            0.236842
48   金榮宇  김영우        51       52     0.980769     0.980769            0.090334
49   金榮珠  김영주        44       44            1     1.00

In [40]:
senators_w_clusters

한글이름          x          y      정당                              소속위원회  \
0    강길부  -7.843739   9.154598     무소속                    산업통상자원중소벤처기업위원회   
1    강병원 -11.518075   6.042161  더불어민주당                            기획재정위원회   
2    강석진   5.722982 -15.073046   미래통합당                      농림축산식품해양수산위원회   
3    강석호   8.049164   6.404850   미래통합당                      농림축산식품해양수산위원회   
4    강창일   3.764485  -7.717409  더불어민주당                            행정안전위원회   
5    강효상  14.575323  -3.458097   미래통합당        공공부문 채용비리 의혹과 관련된 국정조사특별위원회   
6    강훈식   1.413982   2.741007  더불어민주당                            국토교통위원회   
7    경대수   5.061145 -14.301801   미래통합당                      농림축산식품해양수산위원회   
8    고용진   2.713527  -0.616596  더불어민주당                            국회운영위원회   
9    곽대훈   7.267257   0.719697   미래통합당                    산업통상자원중소벤처기업위원회   
10   곽상도  -9.597801   3.696174   미래통합당                              교육위원회   
11   권미혁 -13.270166   6.737703  더불어민주당                            행정안전위원회   
12   권성동  -7.333122   7.424300   미래통합당                            기획재정위원회   
13   권은희 -13.462697   4.140561    국민의당                            행정안전위원회   
14   권칠승  10.766316 -10.074862  더불어민주당                    산업통상자원중소벤처기업위원회   
15   금태섭  -6.218414  -5.078003  더불어민주당  대법관(김선수, 노정희,이동원)임명동의에관한인사청문특별위원회   
16   기동민  -5.329754  -2.087694  더불어민주당                            보건복지위원회   
17   김경협  -9.449069   6.276226  더불어민주당                            기획재정위원회   
18   김관영   5.763693  -5.865829     무소속                    산업통상자원중소벤처기업위원회   
19   김광림  -6.305140   8.931992   미래통합당                            기획재정위원회   
20   김광수   6.864320  -6.003526     민생당                            보건복지위원회   
21   김규환   0.818225   7.244943   미래통합당                    산업통상자원중소벤처기업위원회   
22   김기선  12.898429  -6.342942   미래통합당                    산업통상자원중소벤처기업위원회   
23   김도읍  12.523501  -5.740054   미래통합당  대법관(김선수, 노정희,이동원)임명동의에관한인사청문특별위원회   
24   김동철 -10.139237   5.524635     민생당                            환경노동위원회   
25   김두관  -4.792307   3.483311  더불어민주당                            기획재정위원회   
26   김명연  -4.211778   0.462378   미래통합당                            보건복지위원회   
27   김무성   0.221788  -0.685798   미래통합당                            외교통일위원회   
28   김민기  -0.997969  -5.043008  더불어민주당                            행정안전위원회   
29   김병관   1.063021   0.753651  더불어민주당                            행정안전위원회   
30   김병기   2.458187  -4.335574  더불어민주당                              정보위원회   
31   김병욱   2.269042   0.800899  더불어민주당                              정무위원회   
32   김부겸  -8.432260   5.956553  더불어민주당                            외교통일위원회   
33   김삼화   6.484938  -6.951818   미래통합당                    산업통상자원중소벤처기업위원회   
34   김상훈  -3.505637   6.802902   미래통합당                            국토교통위원회   
35   김상희  -7.521309  -1.771377  더불어민주당                            보건복지위원회   
36   김석기   0.125316   4.393303   미래통합당                            국토교통위원회   
37   김선동  -5.410086   7.811806   미래통합당                              정무위원회   
38   김성식 -14.015994  11.940045     무소속                            기획재정위원회   
39   김성원   0.721415   3.718764   미래통합당                              정무위원회   
40   김성찬   5.603760 -14.034132   미래한국당                      농림축산식품해양수산위원회   
41   김성태   2.679404   9.624476   미래통합당                      과학기술정보방송통신위원회   
42   김성태   2.679348   9.624378   미래통합당                            행정안전위원회   
43   김성환  -9.625093  12.318581  더불어민주당                    산업통상자원중소벤처기업위원회   
44   김세연  -8.790363   1.253936   미래통합당                            보건복지위원회   
45   김수민   6.813972  -7.245066   미래통합당                            여성가족위원회   
46   김순례  -4.760561  -0.010228   미래통합당                            보건복지위원회   
47   김승희   0.914598  -5.916406   미래통합당                            보건복지위원회   
48   김영우 -10.485424   1.630484   미래통합당                            행정안전위원회   
49   김영주  10.306426 -10.937550  더불어민주당                          문화체육관광위원회   
50   김영진   1.108691   2.322298  더불어민주당                            기획재정위원회   


In [48]:
df_merged = senators_w_clusters[['한글이름', '정당','cluster_13']].merge(bills_index[['한글이름', 'bills_index','bills_index_scaled']], on='한글이름', how='left')
df_merged

한글이름      정당  cluster_13  bills_index  bills_index_scaled
0    강길부     무소속           9     0.343284            0.031618
1    강병원  더불어민주당           0     2.473684            0.227839
2    강석진   미래통합당          10     1.297872            0.119541
3    강석호   미래통합당           4     1.340426            0.123460
4    강창일  더불어민주당           1     2.052632            0.189058
5    강효상   미래통합당          11     0.603175            0.055556
6    강훈식  더불어민주당           7     3.411765            0.314241
7    경대수   미래통합당          10     1.558140            0.143513
8    고용진  더불어민주당           7     2.205882            0.203173
9    곽대훈   미래통합당           4     1.608696            0.148169
10   곽상도   미래통합당           2     0.507246            0.046720
11   권미혁  더불어민주당           0     2.540541            0.233997
12   권성동   미래통합당           9     1.018868            0.093843
13   권은희    국민의당           0     1.413793            0.130218
14   권칠승  더불어민주당           8     2.222222            0.204678
15   금태섭  더불어민주당          12     1.151515            0.106061
16   기동민  더불어민주당          12     3.939394            0.362839
17   김경협  더불어민주당           2     2.942857            0.271053
18   김관영     무소속           1     2.843750            0.261924
19   김광림   미래통합당           9     0.810345            0.074637
20   김광수     민생당           1     3.259259            0.300195
21   김규환   미래통합당           6     1.720930            0.158507
22   김기선   미래통합당          11     0.806452            0.074278
23   김도읍   미래통합당          11     1.444444            0.133041
24   김동철     민생당           2     1.914286            0.176316
25   김두관  더불어민주당           2     1.204082            0.110902
26   김명연   미래통합당          12     1.255319            0.115622
27   김무성   미래통합당           7     0.563380            0.051890
28   김민기  더불어민주당           5     2.368421            0.218144
29   김병관  더불어민주당           7     1.363636            0.125598
30   김병기  더불어민주당           1     3.088235            0.284443
31   김병욱  더불어민주당           7     3.117647            0.287152
32   김부겸  더불어민주당           2     0.795455            0.073266
33   김삼화   미래통합당           1     3.600000            0.331579
34   김상훈   미래통합당           6     1.346939            0.124060
35   김상희  더불어민주당          12     4.870968            0.448642
36   김석기   미래통합당           6     2.405405            0.221550
37   김선동   미래통합당           9     1.166667            0.107456
38   김성식     무소속           0     0.193548            0.017827
39   김성원   미래통합당           7     2.857143            0.263158
40   김성찬   미래한국당          10     2.567568            0.236486
41   김성태   미래통합당           6     1.886364            0.173744
42   김성태   미래통합당           6     1.886364            0.173744
43   김성태   미래통합당           6     1.886364            0.173744
44   김성태   미래통합당           6     1.886364            0.173744
45   김성환  더불어민주당           9     0.214286            0.019737
46   김세연   미래통합당           2     1.333333            0.122807
47   김수민   미래통합당           1     3.375000            0.310855
48   김순례   미래통합당          12     1.019608            0.093911
49   김승희   미래통합당           5     2.571429            0.236842
50   김영우   미래통합당           2     0.980769            0.090334
51   김영주  더불어민주당           8     1.000000            0.092105
52   김영진  더불어민주당           7     1.365854            0.125802
53   김영춘  더불어민주당           8     1.375000            0.126645
54   김영호  더불어민주당          12     2.742857            0.252632
55   김용태   미래통합당           9     0.758065            0.069822
56   김재경   미래통합당           7     0.943396            0.086892
57   김재원   미래통합당          11     0.580000            0.053421
58   김정우  더불어민주당           7     4.896552            0.450998
59   김정재   미래통합당          11     2.025641            0.186572
60   김정호  더불어민주당           6     0.750000            0.069079
61   김정훈   미래통합당           9     0.492958            0.045404
62   김종대     정의당           0     2.703704            0.249025
63   김종민  더불어민주당           2     3.0

In [50]:
df_merged.groupby('cluster_13')[['bills_index']].mean()

bills_index
cluster_13             
0              1.681261
1              2.250452
2              1.465430
3              4.385234
4              0.996764
5              1.338190
6              1.658192
7              2.490697
8              2.483238
9              1.262657
10             1.721633
11             1.406662
12             2.374111

In [51]:
df_merged.groupby('cluster_13')[['bills_index_scaled']].mean()

bills_index_scaled
cluster_13                    
0                     0.154853
1                     0.207278
2                     0.134974
3                     0.403903
4                     0.091807
5                     0.123254
6                     0.152728
7                     0.229406
8                     0.228719
9                     0.116297
10                    0.158572
11                    0.129561
12                    0.218668

In [55]:
df_merged['bills_index_scaled'].mean()

0.17130570797659025

In [62]:
df_merged[df_merged['cluster_13']==5]

한글이름      정당  cluster_13  bills_index  bills_index_scaled
28   김민기  더불어민주당           5     2.368421            0.218144
49   김승희   미래통합당           5     2.571429            0.236842
80   나경원   미래통합당           5     0.582090            0.053614
83   도종환  더불어민주당           5     0.871795            0.080297
84   맹성규  더불어민주당           5     1.217391            0.112128
90   박경미  더불어민주당           5     1.681818            0.154904
133  송희경   미래통합당           5     1.750000            0.161184
137  신상진   미래통합당           5     0.912281            0.084026
145  안민석  더불어민주당           5     2.105263            0.193906
151  염동열   미래통합당           5     0.892857            0.082237
166  유은혜  더불어민주당           5     2.459459            0.226529
215  이태규    국민의당           5     0.395833            0.036458
236  전희경   미래통합당           5     0.724138            0.066697
247  정은혜  더불어민주당           5     0.000000            0.000000
258  조승래  더불어민주당           5     2.552632            0.235111
269  채이배     민생당           5     1.724138            0.158802
277  최재성  더불어민주당           5     0.806452            0.074278
283  한선교   미래한국당           5     0.471429            0.043421

전체 입법지수(bills_index_scaled)의 평균을 살펴보면 0.1713이다.  

### cluster 1
cluster 1의 입법지수는 0.15로 전체 평균보다는 낮은 수준이다.  
소속의원: 유승민, 박영선, 오신환, 이혜훈, 이철회 및 정의당 의원 다수  

### cluster 4
cluster 4의 입법지수는 0.403으로 전체보다 2배 이상 높은 것을 확인할 수 있었다.  
소속의원: 이언주, 정동영, 이종걸, 이찬열, 황주홍  

### cluster 6
cluster 6의 입법 지수는 0.1527로 전체 평균보다는 낮은 수준이다.  
소속의원: 나경원, 유은혜, 채이배, 이태규 